src/goto-instrument/contracts/dynamic-frames/dfcc_base_idents.cpp

```diff
@@ -0,0 +1,182 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+#include "dfcc_base_idents.h"
+
+#include <util/byte_operators.h>
+#include <util/cprover_prefix.h>
+#include <util/expr.h>
+#include <util/std_code.h>
+
+#include <iostream>
+
+/// \brief Skips over a typecast expression
+static const exprt &skip_typecast(const exprt &expr)
+{
+  if(expr.id() == ID_typecast)
+  {
+    return skip_typecast(expr.operands().at(0));
+  }
+  return expr;
+}
+
+/// \brief Tries to match an expression of the form `arr[idx]`
+/// where arr has an array type and returns `arr` if successfull
+static optionalt<exprt> match_index_of_array(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_index)
+  {
+    const exprt &op0 = skipped.operands().at(0);
+    if(op0.type().id() == ID_array)
+    {
+      return op0;
+    }
+  }
+  return nullopt;
+}
+
+/// \brief Tries to match \p expr with pattern `arr[idx]` modulo typecasts
+/// where arr has an array type and returns `arr` if successfull
+static optionalt<exprt> match_address_of(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_address_of)
+  {
+    return skipped.operands().at(0);
+  }
+  return nullopt;
+}
+
+/// \brief Tries to match \p expr with pattern `&arr[idx]` modulo typecasts
+/// where `arr` has an array type and returns `arr` if successfull.
+static optionalt<exprt> match_adrress_of_array_index(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_address_of)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    return match_index_of_array(subexpr);
+  }
+  return nullopt;
+}
+
+/// \brief Tries to match \p expr with pattern `*(&subexpr)` modulo typecasts
+/// `subexpr` if successfull
+static optionalt<exprt> match_dereference_of_address_of(const exprt &expr)
+{
+  const exprt &skipped = skip_typecast(expr);
+  if(skipped.id() == ID_dereference)
+  {
+    const exprt &subexpr = skipped.operands().at(0);
+    return match_address_of(subexpr);
+  }
+  return nullopt;
+}
+
+static bool dfcc_base_idents_rec(const exprt &expr, std::set<irep_idt> &dest)
+{
+  // The case-splitting on the lhs copied from symex_assignt::assign_rec
+  // and should cover all cases that symex knows about assignments.
+  if(expr.id() == ID_symbol)
```

Comment:
I believe you could get rid of most of the code (with the exception of `ID_if`) by using `object_descriptor_exprt::root_object(expr)`.